Framework/DataHandling/src/LoadSampleEnvironment.cpp

```diff
@@ -0,0 +1,322 @@
+  MatrixWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+
+  if (inputWS != outputWS) {
+    outputWS = inputWS->clone();
+  }
+
+  const std::string filename = getProperty("Filename");
+  const std::ifstream file(filename.c_str());
+  if (!file) {
+    g_log.error("Unable to open file: " + filename);
+    throw Exception::FileError("Unable to open file: ", filename);
+  }
+
+  boost::shared_ptr<MeshObject> environmentMesh = nullptr;
+
+  std::unique_ptr<LoadAsciiStl> asciiStlReader = nullptr;
+  std::unique_ptr<LoadBinaryStl> binaryStlReader = nullptr;
+
+  if (getProperty("SetMaterial")) {
+    ReadMaterial::MaterialParameters params;
+    params.chemicalSymbol = getPropertyValue("ChemicalFormula");
+    params.atomicNumber = getProperty("AtomicNumber");
+    params.massNumber = getProperty("MassNumber");
+    params.sampleNumberDensity = getProperty("SampleNumberDensity");
+    params.zParameter = getProperty("ZParameter");
+    params.unitCellVolume = getProperty("UnitCellVolume");
+    params.sampleMassDensity = getProperty("SampleMassDensity");
+    params.coherentXSection = getProperty("CoherentXSection");
+    params.incoherentXSection = getProperty("IncoherentXSection");
+    params.attenuationXSection = getProperty("AttenuationXSection");
+    params.scatteringXSection = getProperty("ScatteringXSection");
+    binaryStlReader = std::make_unique<LoadBinaryStl>(filename, params);
+    asciiStlReader = std::make_unique<LoadAsciiStl>(filename, params);
+  } else {
+    binaryStlReader = std::make_unique<LoadBinaryStl>(filename);
+    asciiStlReader = std::make_unique<LoadAsciiStl>(filename);
+  }
+
+  if (binaryStlReader->isBinarySTL(filename)) {
+    environmentMesh = binaryStlReader->readStl();
+  } else if (asciiStlReader->isAsciiSTL(filename)) {
+    environmentMesh = asciiStlReader->readStl();
+  } else {
+    throw Kernel::Exception::ParseError(
+        "Could not read file, did not match either STL Format", filename, 0);
+  }
+  environmentMesh = translate(environmentMesh);
+  environmentMesh = rotate(environmentMesh);
+
+  std::string name = getProperty("EnvironmentName");
+  const bool add = getProperty("Add");
+  Sample &sample = outputWS->mutableSample();
+  std::unique_ptr<Geometry::SampleEnvironment> environment = nullptr;
+  if (add) {
+    environment =
+        std::make_unique<Geometry::SampleEnvironment>(sample.getEnvironment());
+    environment->add(environmentMesh);
+  } else {
+    auto can = boost::make_shared<Container>(environmentMesh);
+    environment = std::make_unique<Geometry::SampleEnvironment>(name, can);
+  }
+  // Put Environment into sample.
+
+  std::string debugString =
+      "Environment has: " + std::to_string(environment->nelements()) +
+      " elements.";
+
+  sample.setEnvironment(std::move(environment));
+
+  auto translatedVertices = environmentMesh->getVertices();
+  int i = 0;
+  for (double vertex : translatedVertices) {
+    i++;
+    g_log.information(std::to_string(vertex));
```

Comment:
This is a lot of detail to pump out at information level. Would this be better at `debug`? It is is also generally better for performance to construct the string and then pass it to the logging functions once as they are quite expensive. We could also check the log level and only do this if the log level is set appropriately?

Reply:
thanks for pointing this out, on decently sized inputs this seems to make it much faster :)